ddtrace/profiling/collector/_traceback.pyx

```diff
@@ -1,22 +1,57 @@
+IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 9):
+    from cpython cimport PyFrameObject
+    from cpython.object cimport PyObject
+    from cpython.ref cimport Py_XDECREF
+
+    cdef extern from "<Python.h>":
+        ctypedef struct PyCodeObject:
+            pass
+        PyObject* PyTuple_GetItem(PyObject *p, Py_ssize_t pos)
+        PyObject* PyDict_GetItem(PyObject *p, PyObject *key)
+        PyCodeObject* PyFrame_GetCode(PyFrameObject* frame)
+        PyFrameObject* PyFrame_GetBack(PyFrameObject* frame)
+
+        IF PY_MINOR_VERSION >= 11:
+            PyObject* PyFrame_GetLocals(PyFrameObject* frame)
+            int PyFrame_GetLineNumber(PyFrameObject* frame)
+            PyObject* PyCode_GetVarnames(PyCodeObject* co)
+
+
 cpdef _extract_class_name(frame):
     # type: (...) -> str
     """Extract class name from a frame, if possible.
 
     :param frame: The frame object.
     """
-    if frame.f_code.co_varnames:
-        argname = frame.f_code.co_varnames[0]
-        try:
-            value = frame.f_locals[argname]
-        except KeyError:
-            return ""
-        try:
-            if argname == "self":
-                return object.__getattribute__(type(value), "__name__")  # use type() and object.__getattribute__ to avoid side-effects
-            if argname == "cls":
-                return object.__getattribute__(value, "__name__")
-        except AttributeError:
-            return ""
+    # Python 3.11 moved PyFrameObject to internal C API and cannot be directly accessed from tstate
+    IF PY_MAJOR_VERSION > 3 or (PY_MAJOR_VERSION == 3 and PY_MINOR_VERSION >= 11):
+        code = PyFrame_GetCode(<PyFrameObject*> frame)
+        co_varnames = PyCode_GetVarnames(code)
+        if co_varnames:
+            argname = PyTuple_GetItem(co_varnames, 0)
+            try:
+                f_locals = PyFrame_GetLocals(<PyFrameObject*>frame)
+                value = <object>PyDict_GetItem(f_locals, argname)
```

Comment:
According to https://docs.python.org/3/c-api/dict.html#c.PyDict_GetItem, this returns a borrowed reference. Am I correct in not needing to call `Py_XDECREF(value)` later?

Reply:
Cython is probably INCREF and DECREFing it anyways? I think when we looked at the annotation it was doing that.